File: embulk-core/src/main/java/org/embulk/spi/FileInput.java
package org.embulk.spi;

import java.util.Optional;

public interface FileInput extends AutoCloseable {
    boolean nextFile();

    Buffer poll();

    void close();

    // Gets a text that hints the name of the current file input.
    //
    // <p>The hint is aimed for logging, not for any data recorded in rows / columns.
    // There is no any guarantee on the text. The text format may change in future versions.
    // The text may be lost by putting another plugin in the configuration.
    //
    // @return the hint text
    default Optional<String> hintOfCurrentInputFileNameForLogging() {
        return Optional.empty();
    }
}

File: lib/embulk/file_input.rb

module Embulk
  require 'embulk/buffer'

  class FileInput
    def initialize(java_file_input)
      @java_file_input = java_file_input
    end

    def next_file
      if @java_file_input.nextFile
        return self
      else
        return nil
      end
    end

    def each(&block)
      while java_buffer = @java_file_input.poll
        buffer = Buffer.from_java(java_buffer)
        java_buffer.release
        yield buffer
      end
    end

    def close
      @java_file_input.close
    end
  end
end
